scrapy/utils/conf.py

```diff
@@ -1,36 +1,54 @@
 import os
 import sys
+import warnings
 from operator import itemgetter
 
 import six
 from six.moves.configparser import SafeConfigParser
 
+from scrapy.exceptions import ScrapyDeprecationWarning
+from scrapy.settings import BaseSettings
 from scrapy.utils.deprecate import update_classpath
 
 
-def build_component_list(base, custom, convert=update_classpath):
-    """Compose a component list based on a custom and base dict of components
-    (typically middlewares or extensions), unless custom is already a list, in
-    which case it's returned.
-    """
+def build_component_list(compdict, custom=None, convert=update_classpath):
+    """Compose a component list from a { class: order } dictionary."""
 
     def _check_components(complist):
         if len({convert(c) for c in complist}) != len(complist):
             raise ValueError('Some paths in {!r} convert to the same object, '
                              'please update your settings'.format(complist))
 
-    if isinstance(custom, (list, tuple)):
-        _check_components(custom)
-        return type(custom)(convert(c) for c in custom)
-
     def _map_keys(compdict):
         _check_components(compdict)
-        return {convert(k): v for k, v in six.iteritems(compdict)}
-
-    compdict = _map_keys(base)
-    compdict.update(_map_keys(custom))
-    items = (x for x in six.iteritems(compdict) if x[1] is not None)
-    return [x[0] for x in sorted(items, key=itemgetter(1))]
+        if isinstance(compdict, BaseSettings):
+            compbs = BaseSettings()
+            for k, v in six.iteritems(compdict):
+                prio = compdict.getpriority(k)
+                compbs.set(convert(k), v, priority=prio)
+            return compbs
+        else:
+            return {convert(k): v for k, v in six.iteritems(compdict)}
+
+    # Begin backwards support for (base, custom) call signature
+    if custom is not None:
+        warnings.warn(('_BASE settings are deprecated and calling '
+                       'build_component_list() with the (base, custom) '
+                       'call signature are deprecated.'),
+                      category=ScrapyDeprecationWarning, stacklevel=2)
+        if isinstance(custom, (list, tuple)):
+            _check_components(custom)
+            return type(custom)(convert(c) for c in custom)
+        compdict = BaseSettings(_map_keys(compdict), priority='default')
+        compdict.update(_map_keys(custom))
+    # End backwards support
```

Comment:
I think we should remove support for the old signature, `build_component_list` is an internal helper for scrapy, there's no need to maintain backward compatibility for it and it makes this function slightly more difficult to understand.